ipfs-cluster-ctl/graph.go

```diff
@@ -0,0 +1,238 @@
+func (dW *dotWriter) writeNode(id string, nT nodeType) error {
+	var nodeStr string
+	switch nT {
+	case tCluster:
+		nodeStr = fmt.Sprintf("C%d [label=\"%s\" color=\"blue2\"]\n", len(dW.clusterOrder), shortId(id))
+		dW.clusterOrder[id] = len(dW.clusterOrder)
+	case tIpfs:
+		nodeStr = fmt.Sprintf("I%d [label=\"%s\" color=\"goldenrod\"]\n", len(dW.ipfsOrder), shortId(id))
+		dW.ipfsOrder[id] = len(dW.ipfsOrder)
+	default:
+		return errUnknownNodeType
+	}
+	n, err := io.WriteString(dW.w, nodeStr)
+	if err == nil && n != len([]byte(nodeStr)) {
+		err = errUnfinishedWrite
+	}
+	return err
+}
+
+func (dW *dotWriter) print() error {
+	_, err := io.WriteString(dW.w, "digraph cluster {\n\n")
+	err = dW.writeComment("The nodes of the connectivity graph")
+	if err != nil {
+		return err
+	}
+	err = dW.writeComment("The cluster-service peers")
+	if err != nil {
+		return err
+	}
+	// Write cluster nodes
+	for k := range dW.clusterEdges {
+		err = dW.writeNode(k, tCluster)
+		if err != nil {
+			return err
+		}
+	}
+	_, err = io.WriteString(dW.w, "\n")
+	if err != nil {
+		return err
+	}
+	
+	err = dW.writeComment("The ipfs peers")
+	if err != nil {
+		return err
+	}
+	// Write ipfs nodes
+	for k := range dW.ipfsEdges {
+		err = dW.writeNode(k, tIpfs)
+		if err != nil {
+			return err
+		}
+	}
+	_, err = io.WriteString(dW.w, "\n")
+	if err != nil {
+		return err
+	}
+
+	err = dW.writeComment("Edges representing active connections in the cluster")
+	if err != nil {
+		return err
+	}
+	err = dW.writeComment("The connections among cluster-service peers")
+	// Write cluster edges
+	for k, v := range dW.clusterEdges {
+		for _, id := range v {
+			err = dW.writeEdge(k,tCluster,id,tCluster)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	_, err = io.WriteString(dW.w, "\n")
+	if err != nil {
+		return err
+	}
+
+	err = dW.writeComment("The connections between cluster peers and their ipfs daemons")
+	if err != nil {
+		return err
+	}
+	// Write cluster to ipfs edges
+	for k, id := range dW.clusterIpfsEdges {
+		err = dW.writeEdge(k,tCluster,id,tIpfs)
+		if err != nil {
+			return err
+		}
+	}
+	_, err = io.WriteString(dW.w, "\n")
+	if err != nil {
+		return err
+	}
+
+	err = dW.writeComment("The swarm peer connections among ipfs daemons in the cluster")
+	if err != nil {
+		return err
+	}	
+	// Write ipfs edges
+	for k, v := range dW.ipfsEdges {
+		for _, id := range v {
+			err = dW.writeEdge(k,tIpfs,id,tIpfs)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	_, err = io.WriteString(dW.w, "\n }")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// truncate the provided peer id string to the 3 last characters.  Odds of 
+// pairwise collisions are less than 1 in 200,000 so with 70 cluster peers  
+// the chances of a collision are still less than 1 in 100 (birthday paradox).
+// As clusters grow bigger than this we can provide a flag for including
+// more characters.
+func shortId(peerString string) string {
```

Comment:
It probably make sense to use the same short peer identifiers as libp2p. when doing `String() on a peer it returns something like `Peer <abcdef>` where the original peer name was `Qmabcdef....`